### ISSUE TITLE
Precompile contract modexp returns incorrect value in corner case

### ISSUE THREAD

**Issue by last-las** [ethereumjs/ethereumjs-monorepo#3168]
# Bug description

The precompile modexp is bias from the standard impementation.

When the `Bsize`  is zero, [the TypeScript EVM returns a BINGINT_0 directly](https://github.com/ethereumjs/ethereumjs-monorepo/blob/0170906f2880e01951bc56cc2c86fc263da21cc3/packages/evm/src/precompiles/05-modexp.ts#L154-L159), failed to account for the scenario of  $0^0$. In this case, the modexp is expected to return 1 instead of 0.

Furthermore, if the `Msize` is 0, the return value should be empty instead of 0. 

The following bytecode invokes the modexp with `Bsize = 0, Esize = 0, Msize = 1`, the expected output should be `01ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff`, however  the TypeScript EVM returns `00ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff`: 

```
60006000526000602052600160405260ff60605360ee60615360dd6062537fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff608052602060806080600060055afa60206080f3
```

Another example is when the modexp is invoked with `Bsize = 0, Esize = 0, Msize = 0`. Geth returns `ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff`, while the TypeScript EVM returns `00ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff`:

```
60006000526000602052600060405260ff60605360ee60615360dd6062537fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff608052602060806080600060055afa60206080f3
```

**Comment by jochem-brouwer** [comment]
Do you have example calldata for those?

**Comment by jochem-brouwer** [comment]
Can you explain how you got this code?

`60006000526000602052600160405260ff60605360ee60615360dd6062537fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff608052602060806080600060055afa60206080f3`

It fills the end of the memory block with 0xffff..ffff. The actual output of the precompile is 0x00, infura reports 0x01 (so it looks like we have a problem indeed)

**Comment by jochem-brouwer** [comment]
We can invoke this via infura:

```
curl https://mainnet.infura.io/v3/<YOUR_API_KEY> \
  -X POST \
  -H "Content-Type: application/json" \
  -d '{"jsonrpc":"2.0","method":"eth_call","params": [{"gas": "0xffffff","data": "0x60006000526000602052600160405260ff60605360ee60615360dd6062537fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff608052602060806080600060055afa60206080f3"}, "latest"],"id":1}'
```

Infura: `0x01ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff`, We: `00ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff`

For: `60006000526000602052600060405260ff60605360ee60615360dd6062537fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff608052602060806080600060055afa60206080f3`

Infura: `0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff`, we: `00ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff`

**Comment by last-las** [comment]
> Can you explain how you got this code?
> 
> `60006000526000602052600160405260ff60605360ee60615360dd6062537fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff608052602060806080600060055afa60206080f3`
> 
> It fills the end of the memory block with 0xffff..ffff. The actual output of the precompile is 0x00, infura reports 0x01 (so it looks like we have a problem indeed)

We detect it using a fuzzer and handcraft the bytecode here

**Comment by jochem-brouwer** [comment]
Very nice, is this fuzzer open source? Would love to see it in action :smile:

**Comment by last-las** [comment]
> Very nice, is this fuzzer open source? Would love to see it in action 😄

sry but currently it's not :)

**Comment by holgerd77** [comment]
This will be fixed in #3261
